File: API/Entities/Basket.cs
namespace API.Entities;

public class Basket
{
    public int Id { get; set; }

    //
    // this is used to stored id of basket as cookie
    //
    public required string BasketId { get; set; }

    // 1-* relationship (i.e) one basket can have many items
    public List<BasketItem> Items { get; set; } = [];

    public void AddItem(Product product, int quantity)
    {
        if (product == null)
            ArgumentNullException.ThrowIfNull(product);

        if (quantity <= 0)
            throw new ArgumentException("Quantity must be greater than zero");

        var existingItem = FindItem(product.Id);

        if (existingItem != null)
        {
            existingItem.Quantity += quantity;
        }
        else
        {
            Items.Add(new BasketItem
            {
                Product = product,
                Quantity = quantity,
            });
        }
    }

    public void RemoveItem(int productId, int quantity)
    {
        if (quantity <= 0)
            throw new ArgumentException("Quantity must be greater than zero");

        var existingItem = FindItem(productId);

        if (existingItem == null) return;

        existingItem.Quantity -= quantity;

        if (existingItem.Quantity <= 0)
            Items.Remove(existingItem);
    }

    private BasketItem? FindItem(int productId)
    {
        return Items.FirstOrDefault(x => x.ProductId == productId);
    }
}

File: react-client/src/app/models/basket.ts
import {Product} from "./product.ts";

export type Basket ={
    id: number
    basketId: string
    items: Item[]
}

export class Item{
    
    constructor(product:Product, quantity:number){
        this.productId = product.id;
        this.name = product.name;
        this.quantity = quantity;
        this.price=product.price;
        this.pictureUrl= product.pictureUrl;
        this.type = product.type;
        this.brand= product.brand;
    }
    
    productId: number
    name: string
    price: number
    pictureUrl: string
    brand: string
    type: string
    quantity: number
}